library/src/main/java/com/datatorrent/lib/join/AbstractManagedStateInnerJoinOperator.java

```diff
@@ -0,0 +1,265 @@
+    long timeBucket = extractTime(tuple,isStream1Data);
+    if (!((ManagedTimeStateMultiValue)store).put(key, tuple,timeBucket)) {
+      return;
+    }
+    Spillable.SpillableByteArrayListMultimap<K, T> valuestore = isStream1Data ? stream2Data : stream1Data;
+    Future<List> future = ((ManagedTimeStateMultiValue)valuestore).getAsync(key);
+    if (future.isDone()) {
+      try {
+        joinStream(tuple,isStream1Data, future.get());
+      } catch (InterruptedException e) {
+        throw new RuntimeException(e);
+      } catch (ExecutionException e) {
+        throw new RuntimeException(e);
+      }
+    } else {
+      waitingEvents.put(new JoinEvent<>(key,tuple,isStream1Data),future);
+    }
+  }
+
+  @Override
+  public void handleIdleTime()
+  {
+    if (waitingEvents.size() > 0) {
+      processWaitEvents();
+    } else {
+      /* nothing to do here, so sleep for a while to avoid busy loop */
+      try {
+        Thread.sleep(sleepMillis);
+      } catch (InterruptedException ie) {
+        throw new RuntimeException(ie);
+      }
+    }
+  }
+
+  @Override
+  public void beforeCheckpoint(long l)
+  {
+    stream1Store.beforeCheckpoint(l);
+    stream2Store.beforeCheckpoint(l);
+  }
+
+  @Override
+  public void checkpointed(long l)
+  {
+    stream1Store.checkpointed(l);
+    stream2Store.checkpointed(l);
+  }
+
+  @Override
+  public void committed(long l)
+  {
+    stream1Store.committed(l);
+    stream2Store.committed(l);
+  }
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    sleepMillis = context.getValue(Context.OperatorContext.SPIN_MILLIS);
+    super.setup(context);
+    ((FileAccessFSImpl)stream1Store.getFileAccess()).setBasePath(context.getValue(DAG.APPLICATION_PATH) + Path.SEPARATOR + stateDir + Path.SEPARATOR + String.valueOf(context.getId()) + Path.SEPARATOR + stream1State);
+    ((FileAccessFSImpl)stream2Store.getFileAccess()).setBasePath(context.getValue(DAG.APPLICATION_PATH) + Path.SEPARATOR + stateDir + Path.SEPARATOR + String.valueOf(context.getId()) + Path.SEPARATOR + stream2State);
+    stream1Store.getCheckpointManager().setRecoveryPath("managed_state_" + stream1State);
+    stream1Store.getCheckpointManager().setRecoveryPath("managed_state_" + stream2State);
+    stream1Store.setup(context);
+    stream2Store.setup(context);
+  }
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    stream1Store.beginWindow(windowId);
+    stream2Store.beginWindow(windowId);
+    super.beginWindow(windowId);
+  }
+
+  /**
+   * Process the waitingEvents.
+   */
+  private void processWaitEvents()
+  {
+    Iterator<Map.Entry<JoinEvent<K,T>, Future<List>>> waitIterator = waitingEvents.entrySet().iterator();
+    while (waitIterator.hasNext()) {
+      Map.Entry<JoinEvent<K,T>, Future<List>> waitingEvent = waitIterator.next();
+      try {
+        JoinEvent<K,T> event = waitingEvent.getKey();
+        joinStream(event.value,event.isStream1Data,waitingEvent.getValue().get());
```

Comment:
If this Future request has not completed, the Future.get() call will hold the operator thread.
Suggesting to do following:
if (waitingEvent.getValue().isDone()) {
  // Process join
}
else {
  // To make sure idempotency is maintained join needs to happen in the same order
  return;
}